### ISSUE TITLE
Auto Inclination Accuracy Investigation: QZ (IOS) Control of Technogym Myrun

### ISSUE THREAD

**Issue by brumohr** [cagnulein/qdomyos-zwift#610]
**Describe the bug**
A 3 mile Zwift run today on Zwift Fox Hill (London) with a 2 mile climb.  The average gradient for [Fox Hill ](https://veloviewer.com/segment/16802545) is 4.9%.

Observations:
Zwift Inclination		QZ Inclination		Treadmill (follows QZ)
0-4%				tracks nicely			tracks nicely
5%					8.4%				8.4%
6%					10.5%				10.5%
7%					12%					maxed out

When the gradient as shown on Zwift is below 5%, QZ pretty much agrees.

When the gradient as shown on Zwift is 5 to 6%, QZ is about 3.4% higher.

When the gradient as shown on Zwift is 6 to 7%, QZ is about 4.5% higher.

When the gradient as shown on Zwift is above 7%, QZ is about 5% higher.

**To Reproduce**
1. Create Meetup Run on Zwift - London/Fox Hill
2. When on Fox Hill climb monitor Zwift indicated gradient and compare to QZ indicated gradient.
3. See large difference between Zwift and QZ when gradient is above 5%.

**Setup**
 - Zwift running on PC Win 11
 - QZ running on iPad IOS 15.2

**Append a debug log**
Log to be added later today.

**Comment by cagnulein** [comment]
@brumohr yeah please send me the full raw debug log because i need to grep it and with the pdf is impossible. thanks

**Comment by brumohr** [comment]
> @brumohr <https://github.com/brumohr> yeah please send me the full raw debug log because i need to grep it and with the pdf is impossible. thanks
> 

Reference the time indexes on the pdf’s for target capture which will be the most reliable.

**Comment by cagnulein** [comment]
@brumohr 
[inc.txt](https://github.com/cagnulein/qdomyos-zwift/files/7925981/inc.txt)
here you can find the raw value that Zwift send. Could you please analyze it for me? i mean let's check if you see something strange.
If they are correct with the one that you see on Zwift means that i'm doing something wrong
If they are correct with the one that you see on the treadmill means that zwift doesn't send the same inclination that shows (as i guess, infact i guess it shows a sort of average)

Let me know

**Comment by brumohr** [comment]
> @brumohr <https://github.com/brumohr>
> inc.txt <https://github.com/cagnulein/qdomyos-zwift/files/7925981/inc.txt>
> here you can find the raw value that Zwift send. Could you please analyze it for me?
> 

I should probably redo this run on Zwift to be absolutely sure but as I recall Zwift indicated 0% gradient during this brief ramp up before turning onto the climb proper.
QZ displayed an increasing then decreasing gradient which agreed with the terrain as I saw it in game on the monitor.

**Comment by brumohr** [comment]
> inc.txt <https://github.com/cagnulein/qdomyos-zwift/files/7925981/inc.txt>
> here you can find the raw value that Zwift send. Could you please analyze it for me? i mean let's check if you see something strange.
> If they are correct with the one that you see on Zwift means that i'm doing something wrong
> If they are correct with the one that you see on the treadmill means that zwift doesn't send the same inclination that shows (as i guess, infact i guess it shows a sort of average)
> 
Roberto - In my previous post I attached a the “Ramp” portion of the raw data.  What I experienced in-game was not what I recall seeing on the mini map for zwift gradient which always indicated 0%.  The rest of the raw data increases and decreases in keeping with my experience in-game but I cannot speak to the accuracy of the numbers because I don’t understand how the raw data relates to the scaled data in-game.

I would like to re-run this route today while videoing the monitor so we can see exactly what zwift is indicating and compare to the log.  Let me know if you think this would be worthwhile.

**Comment by cagnulein** [comment]
> The rest of the Raw Data you sent me follows the terrain I experienced but I do not know how to interpret the raw data, ie, what does the number 560" represent interns of gradient? How is it scaled?

divided by 100, so 560 is 5.6%

**Comment by cagnulein** [comment]
> I would like to re-run this route today while videoing the monitor so we can see exactly what zwift is indicating and compare to the log. Let me know if you think this would be worthwhile.

i guess it could be enough checking the strava route with the log from QZ. so we can see the instant incline during the run, what do you think?

**Comment by brumohr** [comment]
> i guess it could be enough checking the strava route with the log from QZ. so we can see the instant incline during the run, what do you think?
> 
Yesterday, I played around comparing the Strava route with the log but it is a rather tedious exercise trying to sync the two up.

I’m not really sure I understand what specific data would best help you investigate this further.  If you are applying an offset to what zwift feeds QZ then it seems to me a precise correlation of in-game gradient (as viewed on zwift) to QZ log is needed in order to determine if what we are seeing on Zwift really reflects the actual gradient on the road.  A video of the mini map while in-game may help here…?

I have a great deal of past experience (now retired) interpreting flight logs (in octal) with actual in-flight behavior in order to trouble shoot problems with the star-tracking inertial navigation system (Northrop) on board the SR-71.  So I have a feel for what you are up against when validating the data.

Thoughts?

**Comment by cagnulein** [comment]
> If you are applying an offset to what zwift feeds QZ then it seems to me a precise correlation of in-game gradient

no, i'm just applying what you're seeing in the log. So that's why we should be able to match the Strava route to the QZ one. I know that it's not an easy jo to do and also maybe it will bring us to a closed road (zwift send different data for example :) )

in the full qz log you can find also timestamp for current speed and current incline, so maybe you're able to match it better? if you need help to understand the log i'm here, but i guess they're quite readble. I'm using notepad++ to filter the rows

**Comment by brumohr** [comment]
> i'm just applying what you're seeing in the log. So that's why we should be able to match the Strava route to the QZ one. I know that it's not an easy jo to do and also maybe it will bring us to a closed road (zwift send different data for example :) )
> 
> in the full qz log you can find also timestamp for current speed and current incline, so maybe you're able to match it better?

Ok.  I will do as you suggest and spend some more time to compare the log to the Strava route and let you know if I see anything amiss.  But my quick comparison of the Strava route to the log yesterday suggests that the log agrees with Strava route.  However, this does not explain why Zwift mini map may not accurately reflect the actual gradient we experience in-game...Zwift issue..?

**Comment by cagnulein** [comment]
My thought is that zwift made this by design in order to show an average not a real value because probably a real value will be too variations

**Comment by brumohr** [comment]
> My thought is that zwift made this by design in order to show an average not a real value because probably a real value will be too variations

Makes sense.  I have always thought the zwift mini map is very slow to reflect changes in actual in-game gradient.  I will have some fun playing around with this a bit more and report back but I suspect this is a closed non-issue.

**Comment by brumohr** [comment]
[inc.txt](https://github.com/cagnulein/qdomyos-zwift/files/7928163/inc.txt)

Can you filter the original log to give me both raw incline and distance for the range of time shown in the attached file?

I have been comparing the attached file to the Velo [Viewer](https://veloviewer.com/segment/16802545) profile and something doesn't add up.  Would like to take a closer look.

**Comment by brumohr** [comment]
Incline from time index                                        Average inclination Time Index
                                 08:50:00 to 08:51:00                                                 08:50:00 to 08:51:00
original log                      10.6 to 12                                                          about 7.25 (taken from log)

inc.txt  (raw data)          6.37 to 8.22                                                       about 7.25 (estimated by me)

Distance into run about 2.5 miles.  From the profile on the Velo Viewer the gradient during this 1 minute is around 7.5% and never gets anywhere near 12%.  In the original full log the inclination starts increasing at 08:50:00 from 10.6 to 11.9 at 08:50:22 and then goes to 12% from 08:50:21 to 08:51:00.  Again, the profile for Fox Hill is about 7.5% on average for this minute of climb.

Is the term "inclination_avg\":7.2320224719101125}}" a cumulative average of the entire run? 
I leave it to you to draw conclusions but it looks to me like the treadmill is being commanded by the "current inclination" as shown in the original log.  The raw data shown in inc.txt is more more representative of the actual run profile.

**Comment by cagnulein** [comment]
> Is the term "inclination_avg":7.2320224719101125}}" a cumulative average of the run up to that point?

yes it's a cumulative. do you have the link from the VeloViewer charts?
I need fresh mind for this :)

**Comment by cagnulein** [comment]
@brumohr good news! i guess i got it!
Sun Jan 23 08:47:38 2022 1642956458703 Debug: /Users/cagnulein/qdomyos-zwift/src/virtualtreadmill.cpp void virtualtreadmill::slopeChanged(int16_t) "calculated erg grade 5.385"
Sun Jan 23 08:47:38 2022 1642956458712 Debug: /Users/cagnulein/qdomyos-zwift/src/bluetooth.cpp void bluetooth::debug(const QString &) "writing incline 6.27394"

I'm receiving the raw value but i'm applying the Tangent of that value! I'm changing it in the version 2.10.6
Send me an email as a reminder to send you the iOS new version indicating this ticket.
Thanks!

**Comment by brumohr** [comment]
> @brumohr <https://github.com/brumohr> good news! i guess i got it!
> Sun Jan 23 08:47:38 2022 1642956458703 Debug: /Users/cagnulein/qdomyos-zwift/src/virtualtreadmill.cpp void virtualtreadmill::slopeChanged(int16_t) "calculated erg grade 5.385"
> Sun Jan 23 08:47:38 2022 1642956458712 Debug: /Users/cagnulein/qdomyos-zwift/src/bluetooth.cpp void bluetooth::debug(const QString &) "writing incline 6.27394"
> 


Excellent.  Will test today and report back.
Bob.

**Comment by cagnulein** [comment]
Amazing! thanks @brumohr we're a great team together!
